### PR TITLE
Add attachIndex/detachIndex API for Interval Collection Querying

### PR DESCRIPTION
**Pull request by clarenceli-msft**
## Description

Add `attachIndex/detachIndex` API (and related test suites) to make supported indexes customizable

[AB#4152](https://dev.azure.com/fluidframework/internal/_workitems/edit/4152)

### CODE REVIEW

File: packages/dds/sequence/src/intervalCollection.ts [hunk 1]
```diff
@@ -1544,6 +1557,41 @@ export class IntervalCollection<TInterval extends ISerializableInterval> extends
...
+	 */
+	public detachIndex(index: IntervalIndex<TInterval>): boolean {
+		if (!this.attached) {
+			throw new LoggingError("The local interval collection must exist");
+		}
+		for (const interval of this) {
+			index.remove(interval);
+		}
+
+		return this.localCollection?.removeIndex(index) ?? false;
```
Comment (Abe27342) [inline packages/dds/sequence/src/intervalCollection.ts:1592]: if it wasn't present, we should avoid removing intervals from the index. you should also doc the return value here.

File: packages/dds/sequence/src/intervalCollection.ts [hunk 2]
```diff
@@ -1104,6 +1104,19 @@ export class LocalIntervalCollection<TInterval extends ISerializableInterval> {
 		}
 	}
 
+	public appendIndex(index: IntervalIndex<TInterval>) {
+		this.indexes.push(index);
```
[the remark this answers is not in the document]
Comment (clarenceli-msft) [reply packages/dds/sequence/src/intervalCollection.ts:1108]: oh yeah forgot to change that

File: packages/dds/sequence/src/test/intervalCollection.spec.ts [hunk 3]
```diff
@@ -1389,4 +1425,40 @@ describe("SharedString interval collections", () => {
...
+
+			const mockIntervalIndex = new MockIntervalIndex();
+			collection.attachIndex(mockIntervalIndex);
+
+			// Verify if all intervals in this collection are added to the attached index
+			assert.strictEqual(collection.getIntervalById(id1), mockIntervalIndex.get(0));
+			assert.strictEqual(collection.getIntervalById(id2), mockIntervalIndex.get(1));
+
+			// Verify if the intervals in collection are synced with those in collection
+			const id3 = collection.add(2, 5, IntervalType.SlideOnRemove).getIntervalId();
```
Comment (Abe27342) [inline packages/dds/sequence/src/test/intervalCollection.spec.ts:1445]: i would split these assertions into a separate test case. In general, always try to keep your test cases minimal. This has the benefit of having some idea of what went wrong knowing only the test name.

File: packages/dds/sequence/src/intervalCollection.ts (hunk 1 above)
Comment (Abe27342) [reply packages/dds/sequence/src/intervalCollection.ts:1588]: comment still applies on latest

**Review by Abe27342 — COMMENTED** [review]
looks good minus the bit on detachIndex. Might want to test that behavior as well (say an index implementation throws if you try and remove something twice--we should support that kind of index even if someone tries to double-remove it on accident b/c it's not difficult for us to do so and your current implementation doesn't)

File: packages/dds/sequence/src/intervalCollection.ts [hunk 4]
```diff
@@ -1112,6 +1112,10 @@ export class LocalIntervalCollection<TInterval extends ISerializableInterval> {
 		return this.indexes.delete(index);
 	}
 
+	public hasIndex(index: IntervalIndex<TInterval>): boolean {
```
Comment (Abe27342) [inline packages/dds/sequence/src/intervalCollection.ts:1115]: no need to add this API--you can just use the return value of `removeIndex`.

Comment (clarenceli-msft) [reply packages/dds/sequence/src/intervalCollection.ts:1115]: ah yes we can simply reordered the logic. I initially wanted to remove the index after clearing all intervals within it, but it was not necessary.

File: packages/dds/sequence/src/intervalCollection.ts [hunk 5]
```diff
@@ -1544,6 +1552,47 @@ export class IntervalCollection<TInterval extends ISerializableInterval> extends
...
+			index.add(interval);
+		}
+
+		this.localCollection?.appendIndex(index);
+	}
+
+	/**
+	 * Detaches an index from this collection.
+	 * All intervals which are part of this collection will be removed from the index, and updates to this collection
+	 * due to local or remote changes will no longer incur updates to the index.
```
Comment (Abe27342) [inline packages/dds/sequence/src/intervalCollection.ts:1577]: return value should be doc'd here

**Review by Abe27342 — APPROVED** [review]

File: packages/dds/sequence/src/intervalCollection.ts [hunk 6]
```diff
@@ -1104,6 +1104,14 @@ export class LocalIntervalCollection<TInterval extends ISerializableInterval> {
 		}
 	}
 
+	public appendIndex(index: IntervalIndex<TInterval>) {
```
[the remark this answers is not in the document]
Comment (clarenceli-msft) [reply packages/dds/sequence/src/intervalCollection.ts:1107]: yeah thats weird.. btw I am currently working on some other API's in this file, so I can update it manually with future pull requests

**Pull request merged**